### PR TITLE
Restriction rewrite

### LINKED ISSUE #28: Use setfacl for restricting access to apps for specifc user
We can use `setfacl` command line utility to set specific permissions only for a specific user. This should remove the the need for the C kernel communication code and also all the code that's responsible for handling opening a restricted application. Showing the admin or error dialog should be then just handled by Slingshot and Plank themselves. The restricted application won't be accessible from either bash and file manager since the executable will have the right permissions set for that.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/49517684-use-setfacl-for-restricting-access-to-apps-for-specifc-user?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #55: parentcontrol not working
In Juno when I activate parent control for my kids and blocks some apps (appcenter, system settings etc) and when my kids login they can still use app center. system settings... without warning or admin login prompt

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65055052-parentcontrol-not-working?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #75: Don't allow non-admins to see other users' restrictions
I think it's a bit of a privacy concern that non-admin users can see the limits placed on other users without authenticating

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/69679859-don-t-allow-non-admins-to-see-other-users-restrictions?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F60746790&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by donadigo**
This PR overhauls how currently restrictions work within the daemon and how the app restriction works internally. There's a lot of changes but they are needed to progress further and fix #75 and #78.

I would like to mention that this PR does **not** touch UI code at all and so it doesn't fix any issues currently present in the plug code, those will be fixed separately.

Here's what's changed:
1. App restriction was reworked:
* In master we currently have a C library that constantly listens to incoming messages from the kernel and then if a process is launched that is blocked on an account, it is forcefully killed by the daemon. This has a few downsides mainly: 
    - a while loop running and filtering messages constantly
    - the process actually starts but is killed by the daemon which isn't ideal
    - maintaining a C library separately

* In this branch the C library was completely removed and replaced with the following:
   - when an admin blocks an app, the daemon gets a request and uses a built-in `setfacl` (TODO: check if this exists on other distros) command to revoke access to the executable for the specific user e.g:
`setfacl -m u:user:--- /path/to/exec`
This blocks the user from executing the file.

   - To provide the UI when the user wants to launch a restricted exec, the daemon finds the .desktop file for the exec, copies it to the `.local/share/application` directory of the user and replaces the Path field with a path to the PC client that handles showing the not available / authentication dialog (the copying part is done once when the admin sets the restrictions)
This way the app menu or plank don't require any changes.

<s>2. Web restriction command was changed to use URL's instead of IP addresses:
    * In master the daemon has to resolve restricted URL's into IP addresses before sending them off to `iptables`
    * In this branch we just directly send the URL string into `iptables` and let it handle stuff out automatically</s>

3. Logic fixes:
    * Some of the if condititions were just plain wrong like the one here: https://github.com/elementary/switchboard-plug-parental-controls/pull/87/files#diff-87672994b6dd06aa556a9488454c10b1R100, this one was fixed

4. Restriction class simplification:
    * The restriction subclasses now directly receive the `UserConfig` they handle and they can just listen to updates to apply new settings on the fly

5. Admins now get an entry in the config file too:
    * Because of the simplifications in the `Restriction` class, if an account doesn't have an entry in the config file, it will be automatically created so that when an admin restricts itself, the changes can be applied to an already existing config and session

Overall, after this branch, the plug should feel much better and actually work the way it's meant to be.
Currently, master has a lot of problems applying the settings to the current user and handling restrictions for different users correctly.

Fixes #28.
Fixes #55.

### CODE REVIEW

File: src/daemon/UserConfig.vala [hunk 1]
```diff
@@ -23,6 +23,7 @@
 namespace PC.Daemon {
     public class UserConfig : Object {
         public string username { get; set; }
+        public signal void changed (string key);
```
Comment (tintou) [inline src/daemon/UserConfig.vala]: Wouldn't it make more sense to have actual GObject properties and use the property notification system instead? (which is more reliable and gives feedback when a property doesn't exist)

Comment (donadigo) [reply src/daemon/UserConfig.vala]: Indeed but I'm not sure if you can throw errors inside get, set bodies. I will have to test that out.

**Comment by donadigo** [comment]
@danrabbit @tintou updated & resolved.

**Comment by danrabbit** [comment]
Can you split this out into multiple branches? It's a bit too much to review all at once.

Specifically regarding web restrictions, I think we did IP addresses so that it couldn't be circumvented with a proxy, no?

**Comment by donadigo** [comment]
@danrabbit I've removed the changes that affect how web restrictions work and will do a pull request separately. I don't think that I'll be able to revert any more changes in this branch tbh. Master isn't even comparable to this branch because it doesn't really work or it's a hit or miss.

It's now ready to review again.

**Comment by cassidyjames** [comment]
I need to find where this is in the code, but it looks like the icon might be broken in this dialog?

image

**Comment by cassidyjames** [comment]
Somehow when enabling "Allow access to these apps with admin permission", I get many launches of apps after I authenticate. I.e. block Calculator, then try to launch it, authenticate, and 7 copies pop up. Same with a web browser trying to open a link, authenticate, then several windows or tabs open.

File: src/common/Utils.vala [hunk 2]
```diff
@@ -120,6 +118,44 @@ namespace PC {
...
+                        args = exec.split (" ");
+                    }
+
+                    exec = args[0];
+                }
+            }
+
+            return exec_to_path (exec);
+        }
+
```
Comment (jeremypw) [inline src/common/Utils.vala]: Extraneous spaces at end of line 150.

File: src/daemon/AccessControlLists.vala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            if (info.should_show () && info.get_executable () == basename) {
+                return info;
+            }
+        }
+
+        return null;
+    }
+
+    private static File? create_desktop_files_dir (string username) {
+        string local_appdir = Path.build_filename (GLib.Path.DIR_SEPARATOR_S, "home", username, ".local", "share", "applications");
```
Comment (jeremypw) [inline src/daemon/AccessControlLists.vala]: Long line

File: src/daemon/AccessControlLists.vala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            return null;
+        }
+
+        string[] args;
+        string exec = Utils.info_to_exec_path (new DesktopAppInfo.from_keyfile (keyfile), out args);
+
+        if (admin) {
+            args[0] = exec;
+            keyfile.set_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_EXEC,
+                "%s -a \"%s:%s:%s\"".printf (Constants.CLIENT_PATH, username, Constants.PARENTAL_CONTROLS_ACTION_ID, string.joinv (" ", args)));
```
Comment (jeremypw) [inline src/daemon/AccessControlLists.vala]: Long line

File: src/daemon/AccessControlLists.vala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+            var dest = appdir_file.get_child (target_file.get_basename ());
+            if (dest.query_exists ()) {
+                if (permissions_str == DEFAULT_PERMISSIONS) {
+                    try {
+                        dest.delete ();
+                    } catch (Error e) {
+                        warning (e.message);
+                    }
+
```
Comment (jeremypw) [inline src/daemon/AccessControlLists.vala]: Whitespace at end

File: src/daemon/AccessControlLists.vala [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                var os = dest.create (FileCreateFlags.REPLACE_DESTINATION);
+                os.write_all (contents.data, null);
+                os.close ();
+            } catch (Error e) {
+                warning (e.message);
+            }
+        }
+    }
+
+    public static void apply_targets (string username, string[] old_targets, string[] targets, bool old_admin, bool admin) {
```
Comment (jeremypw) [inline src/daemon/AccessControlLists.vala]: Long line.

File: src/daemon/AccessControlLists.vala [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        apply_permissions (username, old_targets, DEFAULT_PERMISSIONS, old_admin);
+        apply_permissions (username, targets, NO_EXEC_PERMISSIONS, admin);
+    }
+}
```
Comment (jeremypw) [inline src/daemon/AccessControlLists.vala]: Missing new line

File: src/daemon/Restrictions/TimeRestriction.vala [hunk 8]
```diff
@@ -21,18 +21,70 @@
...
+                GLib.Source.remove (timeout_id);
+            }
+
+            timeout_ids.clear ();
+            update_timeout_id = 0U;
+        }
+
+        private void update () {
+            stop ();
+            start ();
```
Comment (jeremypw) [inline src/daemon/Restrictions/TimeRestriction.vala]: White space at end of line

File: src/daemon/Server.vala [hunk 9]
```diff
@@ -63,18 +65,62 @@ namespace PC.Daemon {
...
-        public signal void launch (string[] args);
-        public signal void show_app_unavailable (string path);
+        public signal void launch (string[] args, bool incoming);
         public signal void show_timeout (int hours, int minutes);
         public signal void config_changed ();
 
-        public void end_app_authorization (BusName sender) throws GLib.Error {
-            uint32 pid = get_pid_from_sender (sender);
-            app_authorization_ended ((int)pid);
+        public void finish_app_authorization (BusName sender, string username, string[] args) throws GLib.Error, ParentalControlsError {
```
Comment (jeremypw) [inline src/daemon/Server.vala]: Long line.

File: src/daemon/Server.vala [hunk 10]
```diff
@@ -63,18 +65,62 @@ namespace PC.Daemon {
...
+            var config = UserConfig.get_for_username (username, false);
+            if (config == null || !config.admin) {
+                throw new ParentalControlsError.USER_CONFIG_NOT_VAILD ("Error: config for %s is not valid or does not have an ability to authorize".printf (username));
```
Comment (jeremypw) [inline src/daemon/Server.vala]: Long line

File: src/daemon/Server.vala [hunk 11]
```diff
@@ -63,18 +65,62 @@ namespace PC.Daemon {
...
+            }
+
+            uint32 client_pid = get_pid_from_sender (sender);
+            var authority = Polkit.Authority.get_sync ();
+            try {
+                var unix_user = new Polkit.UnixUser.for_name (username);
+                var result = authority.check_authorization_sync (new Polkit.UnixProcess.for_owner ((int)client_pid, 0, unix_user.get_uid ()),
```
Comment (jeremypw) [inline src/daemon/Server.vala]: Long line

File: src/plug/Views/AppsView.vala [hunk 12]
```diff
@@ -165,7 +165,7 @@ namespace PC.Widgets {
 
             string[] targets = {};
             foreach (var entry in entries) {
-                targets += Environment.find_program_in_path (entry.app_info.get_executable ());
+                targets += Utils.info_to_exec_path (entry.app_info, null);
             }
 
             Utils.get_api ().set_user_daemon_targets.begin (user.get_user_name (), targets);
```
Comment (jeremypw) [inline src/plug/Views/AppsView.vala]: White space at end of line

**Comment by mwoyak** [comment]
Sorry guys, but I'm rather new to this so please bear with me.  It seems like the parental control not working issue dates back to at least February of 2019 and when I freshly downloaded the OS earlier this month, it still wasn't working with a native installation of the most up-to-date OS image.  Unfortunately, I don't have the Linux or coding experience to understand how to implement the changes indicated above myself.

But with parental controls as part of the basic feature set of the OS, does anyone know if this issue is being addressed in the official OS release?  Or is it up to each person who downloads the OS to implement the work-arounds on their own?

**Pull request merged**